src/daft-table/src/ops/joins/hash_join.rs

```diff
@@ -1,74 +1,372 @@
+    let lkeys = left.eval_expression_list(left_on)?;
+    let rkeys = right.eval_expression_list(right_on)?;
+
+    let (lkeys, rkeys) = match_types_for_tables(&lkeys, &rkeys)?;
+
+    let (lidx, ridx) = if lkeys.columns.iter().any(|s| s.data_type().is_null())
+        || rkeys.columns.iter().any(|s| s.data_type().is_null())
+    {
+        let l_iter = IndexRange::new(0, lkeys.len() as u64)
+            .map(Some)
+            .chain(repeat(None).take(rkeys.len()));
+        let r_iter = repeat(None)
+            .take(lkeys.len())
+            .chain(IndexRange::new(0, rkeys.len() as u64).map(Some));
+
+        let l_arrow = Box::new(arrow2::array::PrimitiveArray::<u64>::from_trusted_len_iter(
+            l_iter,
         ));
-    }
-    let types_not_match = left
-        .columns
-        .iter()
-        .zip(right.columns.iter())
-        .any(|(l, r)| l.data_type() != r.data_type());
-    if types_not_match {
-        return Err(DaftError::SchemaMismatch(
-            "Types between left and right do not match".to_string(),
+        let r_arrow = Box::new(arrow2::array::PrimitiveArray::<u64>::from_trusted_len_iter(
+            r_iter,
         ));
-    }
-
-    // TODO(Clark): Build the probe table on the smaller table, rather than always building it on the left table.
-    let probe_table = left.to_probe_hash_table()?;
-
-    let r_hashes = right.hash_rows()?;
-    let mut left_idx = vec![];
-    let mut right_idx = vec![];
-    let is_equal = build_multi_array_is_equal(
-        left.columns.as_slice(),
-        right.columns.as_slice(),
-        false,
-        false,
-    )?;
-    for (r_idx, h) in r_hashes.as_arrow().values_iter().enumerate() {
-        if let Some((_, indices)) = probe_table.raw_entry().from_hash(*h, |other| {
-            *h == other.hash && {
-                let l_idx = other.idx;
-                is_equal(l_idx as usize, r_idx)
-            }
-        }) {
-            for l_idx in indices {
-                left_idx.push(*l_idx);
+
+        (
+            UInt64Array::from(("left_indices", l_arrow)).into_series(),
+            UInt64Array::from(("right_indices", r_arrow)).into_series(),
+        )
+    } else {
+        // probe on the smaller table
+        let probe_left = lkeys.len() <= rkeys.len();
+
+        let (lkeys, rkeys) = if probe_left {
+            (&lkeys, &rkeys)
+        } else {
+            (&rkeys, &lkeys)
+        };
+
+        let probe_table = lkeys.to_probe_hash_table()?;
+
+        let r_hashes = rkeys.hash_rows()?;
+
+        let is_equal = build_multi_array_is_equal(
+            lkeys.columns.as_slice(),
+            rkeys.columns.as_slice(),
+            false,
+            false,
+        )?;
+
+        let mut left_idx = vec![];
+        let mut right_idx = vec![];
+
+        let mut l_valid = MutableBitmap::new();
+        let mut r_valid = MutableBitmap::new();
+
+        let mut left_idx_used = vec![false; lkeys.len()];
```

Comment:
its also advisable to use a `MutableBitmap` here since `Vec<bool>` uses 8 bits per entry. Instead of pushing you can `set` the bit.